lib/livebook/file_system/utils.ex

```diff
@@ -0,0 +1,109 @@
+defmodule Livebook.FileSystem.Utils do
+  @moduledoc false
+
+  alias Livebook.FileSystem
+
+  @doc """
+  Asserts that the given path is a directory.
+  """
+  @spec assert_dir_path!(FileSystem.path()) :: :ok
+  def assert_dir_path!(path) do
+    unless dir_path?(path) do
+      raise ArgumentError, "expected a directory path, got: #{inspect(path)}"
+    end
+
+    :ok
+  end
+
+  @doc """
+  Asserts that the given path is a regular file path.
+  """
+  @spec assert_regular_path!(FileSystem.path()) :: :ok
+  def assert_regular_path!(path) do
+    unless regular_path?(path) do
+      raise ArgumentError, "expected a regular file path, got: #{inspect(path)}"
+    end
+
+    :ok
+  end
+
+  @doc """
+  Checks if the given path describes a directory.
+  """
+  @spec dir_path?(FileSystem.path()) :: boolean()
+  def dir_path?(path) do
+    String.ends_with?(path, "/")
+  end
+
+  @doc """
+  Checks if the given path describes a regular file.
+  """
+  @spec regular_path?(FileSystem.path()) :: boolean()
+  def regular_path?(path) do
+    not String.ends_with?(path, "/")
+  end
+
+  @doc """
+  Asserts that the given paths are of the same type.
+  """
+  @spec assert_same_type!(FileSystem.path(), FileSystem.path()) :: :ok
+  def assert_same_type!(path1, path2) do
+    if dir_path?(path1) != dir_path?(path2) do
+      raise ArgumentError,
+            "expected paths of the same type, got: #{inspect(path1)} and #{inspect(path2)}"
+    end
+
+    :ok
+  end
+
+  @doc """
+  Converts the given path into dir path by appending a trailing
+  slash if necessary.
+  """
+  @spec ensure_dir_path(String.t()) :: FileSystem.path()
+  def ensure_dir_path(path) do
+    if String.ends_with?(path, "/") do
+      path
+    else
+      path <> "/"
+    end
+  end
+
+  @doc """
+  Converts the given posix error atom into readable error tuple.
+  """
+  @spec posix_error(atom()) :: {:error, FileSystem.error()}
+  def posix_error(error) do
+    message = error |> :file.format_error() |> List.to_string()
+    {:error, message}
+  end
+
+  @doc """
+  Implements `Livebook.FileSystem.resolve_path` assuming Unix-like
+  path conventions.
+
+  This function assumes absolute paths to have a leading "/"
+  and handles sequences such as "." and "..".
+  """
+  @spec resolve_unix_like_path(FileSystem.path(), String.t()) :: FileSystem.t()
+  def resolve_unix_like_path(dir_path, subject) do
+    assert_dir_path!(dir_path)
+
+    subject =
+      if Path.basename(subject) in [".", ".."] do
+        FileSystem.Utils.ensure_dir_path(subject)
+      else
+        subject
+      end
+
+    absolute_path? = String.starts_with?(subject, "/")
+    path = if absolute_path?, do: subject, else: dir_path <> subject
+    path |> String.split("/") |> expand_parts([]) |> Enum.join("/")
```

Comment:
Are we handling multiple paths like this? `///foo/bar`? I think you should pass `trim: true` to split. But that will require to change the join logic and the ".." handling.

Reply:
What we need is splitting on `/` and then removing all empty strings, except the first and last element. That's my verbose attempt: https://github.com/livebook-dev/livebook/pull/492/commits/4ecd47e45546f46dc4e915f2820518d565835851#diff-6ad2026b7225ccfcf90bba86d1d6b5202ebcd9900faec5bf246efb914d25cd7bR109-R115